pyomo/contrib/piecewise/triangulations.py

```diff
@@ -0,0 +1,682 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright (c) 2008-2024
+#  National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+import itertools
+from types import SimpleNamespace
+from enum import Enum
+from pyomo.common.errors import DeveloperError
+
+
+class Triangulation(Enum):
+    Unknown = 0
+    AssumeValid = 1
+    Delaunay = 2
+    J1 = 3
+    OrderedJ1 = 4
+
+
+def get_unordered_j1_triangulation(points, dimension):
+    points_map, num_pts = _process_points_j1(points, dimension)
+    simplices_list = _get_j1_triangulation(points_map, num_pts - 1, dimension)
+    # for now make a duck-typed thing that superficially looks like an instance of
+    # scipy.spatial.Delaunay (these are NDarrays in the original)
+    triangulation = SimpleNamespace()
+    triangulation.points = list(range(len(simplices_list)))
+    triangulation.simplices = {i: simplices_list[i] for i in triangulation.points}
+    triangulation.coplanar = []
+
+    return triangulation
+
+
+def get_ordered_j1_triangulation(points, dimension):
+    points_map, num_pts = _process_points_j1(points, dimension)
+    if dimension == 2:
+        simplices_list = _get_ordered_j1_triangulation_2d(points_map, num_pts - 1)
+    elif dimension == 3:
+        simplices_list = _get_ordered_j1_triangulation_3d(points_map, num_pts - 1)
+    else:
+        simplices_list = _get_ordered_j1_triangulation_4d_and_above(
+            points_map, num_pts - 1, dimension
+        )
+    triangulation = SimpleNamespace()
+    triangulation.points = list(range(len(simplices_list)))
+    triangulation.simplices = {i: simplices_list[i] for i in triangulation.points}
+    triangulation.coplanar = []
+
+    return triangulation
+
+
+# Does some validation but mostly assumes the user did the right thing
+def _process_points_j1(points, dimension):
+    if not len(points[0]) == dimension:
+        raise ValueError("Points not consistent with specified dimension")
+    num_pts = round(len(points) ** (1 / dimension))
+    if not len(points) == num_pts**dimension:
+        raise ValueError(
+            "'points' must have points forming an n-dimensional grid with straight grid lines and the same odd number of points in each axis"
+        )
+    if not num_pts % 2 == 1:
+        raise ValueError(
+            "'points' must have points forming an n-dimensional grid with straight grid lines and the same odd number of points in each axis"
+        )
```

Comment:
It's a little bit strange. The caller basically takes a nicely ordered multidimensional grid in the form of some cartesian product of ordered lists, flattens it out into a list of points so it fits into the shape of the constructor we want to have, then we unflatten it again to retrieve the original lists. We could allow passing the original lists in the PWLF construction, but that isn't very general.

Reply:
Hnmmm, okay. I'm tempted to leave it as is for now since we're not too worried about efficiency. But we should keep this in mind if we're ever worried about performance (ha!).